Initial commit based on cpp-test-runner

pclausen opened the pull request:
WIP 
Initial commit based on cpp-test-runner

SaschaMann commented on tests/example-syntax-error/expected_results.json:
```diff
@@ -0,0 +1,2 @@
+{"version": 2, "status": "error", "tests": [], "message": "/tmp/example-syntax-error/bob.f90:37:47:\\n\\n   37 |   function hey(statement_THIS_IS_A_SYNTAX_ERROR)\\n      |                                               1\\nError: Symbol ‘statement_this_is_a_syntax_error’ at (1) has no IMPLICIT type\\n/tmp/example-syntax-error/bob.f90:39:45:\\n\\n   39 |     character(len=*), intent(in) :: statemen ...
```
This is hard to read and review manually.

Perhaps you could format this test case and then use a Python or jq script to compare the two json files? That way you avoid tests failing due to slightly different formatting as well, since the structure and content is compared in the test.

SaschaMann replied:
In case it's helpful, here's the Julia CI flow to compare the results generated by the test-runner with expected JSON files: https://github.com/exercism/julia-test-runner/blob/main/.github/workflows/ci.yml

pclausen replied:
Thanks for the suggestion. I simply followed the c++ as template. I will have look at jq and the Julia implementation sometime later this week

ErikSchierboom replied:
@pclausen I've added some commits that use jq for the results.json files.

ErikSchierboom commented on bin/run.sh:
```diff
@@ -0,0 +1,58 @@
+    # build successful, now run test
+    export EXERCISM_FORTRAN_JSON=1
+    ctest -V
+    jq '.' results.json > $results_file
```
Calling jq with just `'.'` as the expression will pretty-print the JSON file. I'm not sure if this is needed in this case, but I thought I'd show you how.

ErikSchierboom commented on bin/run.sh:
```diff
@@ -0,0 +1,58 @@
+
+if [ $ret -ne 0 ]; then
+    message=$(cat "${compilation_errors_file_name}" | tr \\n \; | sed 's/;/\\\\n/g')
+    jq -n --arg m "${message}" '{version: 2, status: "error", tests: [], message: $m}' > $results_file
```
Here I'm using `jq` to output the JSON. It's a bit less error-prone than the `echo` approach and arguably a bit easier to read too :)

ErikSchierboom commented on Dockerfile:
```diff
@@ -0,0 +1,11 @@
+RUN apk add --no-cache coreutils curl jq gfortran libc-dev cmake make python3 git
+
+WORKDIR /opt/test-runner/testlib
+RUN curl -R -O https://raw.githubusercontent.com/pclausen/fortran/main/testlib/CMakeLists.txt
+RUN curl -R -O https://raw.githubusercontent.com/pclausen/fortran/main/testlib/TesterMain.f90
```
@pclausen The test runner runs without networking, which means that the testing library can't be downloaded on the fly when running the tests. The solution is to pre-download the testing library within the Dockerfile, which does have access to the network when being built.

pclausen replied:
ahh, did not know that.

ErikSchierboom commented on Dockerfile:
```diff
@@ -0,0 +1,11 @@
+FROM alpine:3.11
+
+RUN apk add --no-cache coreutils curl jq gfortran libc-dev cmake make python3 git
```
I've added `jq` and `curl` as dependencies. `curl` is used below to download the testing library and `jq` is used later to write a nicely formatted `results.json` file.

ErikSchierboom commented on bin/run.sh:
```diff
@@ -0,0 +1,64 @@
+# the makefile uses the directory name to determine the files
+cp -R "${input_dir}/" "${build_dir}" && cd "${build_dir}"
+
+# Copy the testlib which outputs a results.json file when running the tests
+cp -R "${testlib_dir}" .
```
@pclausen This makes the cmake command not download the testing library as it will already be present.

ErikSchierboom commented on tests/example-all-fail/CMakeLists.txt:
```diff
@@ -0,0 +1,71 @@
+# Get the exercise name from the current directory
+get_filename_component(exercise ${CMAKE_CURRENT_SOURCE_DIR} NAME)
```
The `CMakeLists.txt` files had replaced `${CMAKE_CURRENT_SOURCE_DIR}` with `bob`. I've reverted this to better simulate what happens with actual exercises.

ErikSchierboom commented on bin/run.sh:
```diff
@@ -0,0 +1,64 @@
+else
+    # build successful, now run test
+    export EXERCISM_FORTRAN_JSON=1
+    ctest -V
+    jq '.' "${build_results_file}" > "${results_file}"
```
While trying to check the `results.json` output, I'm finding it to be empty. I've run `./bin/run-in-docker.sh example-success $PWD/tests/example-success $PWD/tests/example-success` to run it on the `example-success` test exercise, which outputs:

```
UpdateCTestConfiguration  from :/tmp/example-success/DartConfiguration.tcl
Parse Config file:/tmp/example-success/DartConfiguration.tcl
UpdateCTestConfiguration  from :/tmp/example-success/DartConfiguration.tcl
Parse Config file:/tmp/example-success/DartConfiguration.tcl
Test project /tmp/example-success
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 1
    Start 1: example-success_test

1: Test command: /tmp/example-success/example-success
1: Test timeout computed to be: 1500
1:  Test summary: 0 of 23 tests failed
1: STOP 0
1/1 Test #1: example-success_test .............   Passed    0.00 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   0.00 sec
```

That looks good to me, but the `results.json` in the build directory is empty. It could be due to me messing with the testlib?

ErikSchierboom commented:
@pclausen Thanks for working on this! It looks like the bulk of it is implemented, but I can't get the tests passing on my machine. See the above comments for more information.

pclausen commented:
@ErikSchierboom I think I have now fixed the build and tests. The general idea now is to download CMakeLists.txt and testlib at creation of docker image. I had some strange issues with different quote-signs on docker vs local build and also that line numbers on in the make files and thus in the build errors did not match. So I added removal of quotation-marks and numbers from the output.

pclausen commented:
@ErikSchierboom I was also confused by the 
```
results_file="${output_dir}/results.json"
build_results_file="${build_dir}/results.json"
```
but it seems also to be a construct from the cpp track. My preference would be NOT to copy the results.json file back to `$root` or `$output_dir`, but it is unclear to me that this is by design or not.

ErikSchierboom commented:
> but it seems also to be a construct from the cpp track. My preference would be NOT to copy the results.json file back to $root or $output_dir, but it is unclear to me that this is by design or not.

That is indeed by design, as the [test runner spec](https://github.com/exercism/docs/blob/main/building/tooling/test-runners/interface.md#execution) requires the `results.json` file to be written in `$output_dir`.

ErikSchierboom commented:
@pclausen If I ran `./bin/run-tests-in-docker.sh`, I still got several build failures. Turned out this had to do with old results not being cleaned up, which I've fixed and pushed.

ErikSchierboom reviewed (approved):
This is looking good! I've just tested with the `allergies` exercise and that seemed to work.

The only thing remaining is to add the github codeowners and workflows. For that, you should be able to just copy the contents of https://github.com/exercism/generic-test-runner/tree/main/.github.

pclausen commented:
> This is looking good! I've just tested with the `allergies` exercise and that seemed to work.

Great


> The only thing remaining is to add the github codeowners and workflows. For that, you should be able to just copy the contents of https://github.com/exercism/generic-test-runner/tree/main/.github.

Not quite sure I understand this step but I will open a separate PR for that.

pclausen commented:
@ErikSchierboom I am a bit lost in how to 

> For that, you should be able to just copy the contents of https://github.com/exercism/generic-test-runner/tree/main/.github.

I have tried to add these files in a new branch 
```
workspace/fortran-test-runner $ git diff main --name-only
.github/CODEOWNERS
.github/dependabot.yml
.github/labels.yml
.github/workflows/ci.yml
.github/workflows/docker.yml
.github/workflows/sync-labels.yml
``` 

but I keep getting this error

```
git push --set-upstream origin add-workflow
Enumerating objects: 11, done.
Counting objects: 100% (11/11), done.
Delta compression using up to 16 threads
Compressing objects: 100% (9/9), done.
Writing objects: 100% (10/10), 3.67 KiB | 3.67 MiB/s, done.
Total 10 (delta 1), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (1/1), completed with 1 local object.
To https://github.com/pclausen/fortran-test-runner.git
 ! [remote rejected] add-workflow -> add-workflow (refusing to allow an OAuth App to create or update workflow `.github/workflows/ci.yml` without `workflow` scope)
error: failed to push some refs to 'https://github.com/pclausen/fortran-test-runner.git'
```

Can you please do this change because I am at the end of my github knowledge. Thanks.

iHiD commented:
@pclausen Erik's not around today. In the interim, it looks like your local github token is maybe out of date. See https://stackoverflow.com/questions/64059610/how-to-resolve-refusing-to-allow-an-oauth-app-to-create-or-update-workflow-on for some other people who have solved this problem.

ErikSchierboom commented:
@pclausen I think Jeremy's suggestion is correct. Do you want to take a stab at fixing it using the above link or would you prefer me to do it?
